### PR TITLE
add dataLoader to master

### PR DESCRIPTION
**Pull request by rameziophobia**

### CODE REVIEW

File: data/wild_hands.py [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 from PIL import Image
 from torch.utils.data import Dataset
+from torch.utils.data.dataloader import DataLoader
+from torchvision import transforms
+import matplotlib.pyplot as plt
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: Testing imports should be inside the testing module only (```__main__``` in this case)

File: data/wild_hands.py [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 from PIL import Image
 from torch.utils.data import Dataset
+from torch.utils.data.dataloader import DataLoader
+from torchvision import transforms
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: Transforms should not be imported in the Dataset Class, only when its used outside the module

File: data/wild_hands.py (hunk 1 above)
Comment (MohamedAliRashad) [inline data/wild_hands.py]: No need for ```matplotlib``` for viewing the images, ```PIL``` was enough.

File: data/wild_hands.py [hunk 3]
```diff
@@ -13,11 +16,21 @@ class WildHandsDataset(Dataset):
         target_transform (callable, optional): A function/transform that takes in the
             target and transforms it.
     """
-    def __init__(self, dataset_path, transform=None, target_transform=None):
-        pass
+
+    def __init__(self, dataset_path, level=0, transform=None, target_transform=None):
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: Make the default value for ```level``` is 1, because it can only operate in one of the three levels.

File: data/wild_hands.py [hunk 4]
```diff
@@ -13,11 +16,21 @@ class WildHandsDataset(Dataset):
...
+        if level:
+            self.dataset_path = os.path.join(dataset_path, f"level_{level}")
+        else:
+            self.dataset_path = dataset_path
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: No need for if conditions in the initialization.

File: data/wild_hands.py [hunk 5]
```diff
@@ -13,11 +16,21 @@ class WildHandsDataset(Dataset):
...
+        default_transform = transforms.ToTensor()
+        self.transform = transform if transform is not None else default_transform
+        self.target_transform = target_transform if target_transform is not None else default_transform
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: it's unadvised to assume that the user will need ```ToTensor()``` transform as a default tensor, A better way is to let him choose with the ```transform``` argument.

File: data/wild_hands.py [hunk 6]
```diff
@@ -13,11 +16,21 @@ class WildHandsDataset(Dataset):
...
+        self.img_dirs, self.label_dirs = self.get_img_dirs()
+        self.length = len(self.img_dirs)
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: Not advised to add a variable for length you will only use once in ```__len__```, A better way is to calculate the length directly in the function.

File: data/wild_hands.py [hunk 7]
```diff
@@ -13,11 +16,21 @@ class WildHandsDataset(Dataset):
...
 
     def __len__(self):
-        pass
+        return self.length
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: ```len(self.img_dirs)``` can be a better way.

File: data/wild_hands.py [hunk 8]
```diff
@@ -26,8 +39,35 @@ def __getitem__(self, idx):
         Returns:
             tuple: (image, target) where target is the same image.
         """
-        pass
+        img = Image.open(self.img_dirs[idx]).convert("RGB")
+        img = self.transform(img)
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: Check if there is transforms first.

File: data/wild_hands.py [hunk 9]
```diff
@@ -26,8 +39,35 @@ def __getitem__(self, idx):
...
+
+        target = Image.open(self.label_dirs[idx]).convert("RGB")  # not sure mn el convert tbh
+        target = self.transform(target)
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: Check if there is transforms first, and use ```target_transform``` instead.

File: data/wild_hands.py [hunk 10]
```diff
@@ -26,8 +39,35 @@ def __getitem__(self, idx):
...
+        return img, target
+
+    def get_img_dirs(self):
+        imgs = []
+        labels = []
+        for dir_path, sub_dirs, files in os.walk(self.dataset_path):
+            for file in files:
```
Comment (MohamedAliRashad) [inline data/wild_hands.py:45]: A better way is to assume you will be operating on a certain level so scrabbing can be easier.

File: data/wild_hands.py [hunk 11]
```diff
@@ -26,8 +39,35 @@ def __getitem__(self, idx):
...
+                if 'ask' in file:
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: In labels, you should check if the format is in ```.png```.

File: data/wild_hands.py [hunk 12]
```diff
@@ -26,8 +39,35 @@ def __getitem__(self, idx):
...
+                    labels.append(os.path.join(dir_path, file))
+                else:
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: An else statement can be unsafe in some conditions, Check also if the file has an image format you know (except ```png```).

**Review by MohamedAliRashad — COMMENTED** [review]
Doing Great, Keep up the good work ^^

File: data/wild_hands.py (hunk 10 above)
Comment (rameziophobia) [reply data/wild_hands.py:45]: scrabbing ?

File: data/wild_hands.py (hunk 11 above)
Comment (rameziophobia) [reply data/wild_hands.py]: why ?

File: data/wild_hands.py (hunk 12 above)
Comment (rameziophobia) [reply data/wild_hands.py]: the img frames also end with .png

File: data/wild_hands.py (hunk 10 above)
Comment (MohamedAliRashad) [reply data/wild_hands.py:45]: Getting the images from folders.

File: data/wild_hands.py (hunk 11 above)
Comment (MohamedAliRashad) [reply data/wild_hands.py]: it's an easier way and will allow future support for the class if we added new data.

File: data/wild_hands.py (hunk 12 above)
Comment (MohamedAliRashad) [reply data/wild_hands.py]: I rechecked the dataset and found you were correct, check if it's ```png``` also.

Comment (rameziophobia) [reply data/wild_hands.py]: now I check if the extension is either jpg or png for both the labels and the images

Comment (rameziophobia) [reply data/wild_hands.py]: because level 3 images are jpg

Comment (MohamedAliRashad) [reply data/wild_hands.py]: Right

File: data/wild_hands.py [hunk 13]
```diff
@@ -43,10 +43,11 @@ def get_img_dirs(self):
         labels = []
         for dir_path, sub_dirs, files in os.walk(self.dataset_path):
             for file in files:
-                if 'ask' in file and (os.path.splitext(file)[1]).lower() == '.png':
-                    labels.append(os.path.join(dir_path, file))
-                else:
-                    imgs.append(os.path.join(dir_path, file))
+                if (os.path.splitext(file)[1]).lower() in ['.png', '.jpg']:
```
Comment (MohamedAliRashad) [inline data/wild_hands.py]: use ```endswith()``` instead
```
file.lower().endswith(('.png', '.jpg', '.jpeg'))
```

**Review by MohamedAliRashad — APPROVED** [review]
I think we are good for merging
